src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                    }
+                }
+
+                if (TryConvertQueryExpression(source, internalNodeMethod, out StatementSyntax newNode))
+                {
+                    documentUpdateInfo = new DocumentUpdateInfo(forEachStatement, newNode.WithAdditionalAnnotations(Simplifier.Annotation));
+                    return true;
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertIfInReturnStatement(QueryExpressionSyntax source, ReturnStatementSyntax returnStatement, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // before: return from a in b select a;
+                // after: 
+                // foreach(var a in b)
+                // {
+                //      yield return a;
+                // }
+                StatementSyntax internalNodeMethod(ExpressionSyntax expression)
+                    => SyntaxFactory.YieldStatement(SyntaxKind.YieldReturnStatement, expression);
+
+                if (TryConvertQueryExpression(source, internalNodeMethod, out var newNode))
+                {
+                    documentUpdateInfo = new DocumentUpdateInfo(returnStatement, newNode.WithAdditionalAnnotations(Simplifier.Annotation));
+                    return true;
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertQueryExpression(
+                QueryExpressionSyntax source,
+                Func<ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod,
+                out StatementSyntax statement)
+            {
+                ProcessingState processingState = new ProcessingState(leafExpressionCreationMethod);
+                if (TryProcessFromClause(source.FromClause, processingState) &&
+                    TryProcessQueryBody(source.Body, processingState))
+                {
+                    // Executes syntax building methods from bottom to the top of the tree.
+                    statement = processingState.ExecuteMethods().WithAdditionalAnnotations(Simplifier.Annotation);
+                    return true;
+                }
+
+                statement = default;
+                return false;
+            }
+
+            private bool TryProcessQueryBody(
+                QueryBodySyntax queryBody,
+                ProcessingState processingState)
+            {
+                foreach (var queryClause in queryBody.Clauses)
+                {
+                    if (!TryProcessQueryClause(queryClause, processingState))
+                    {
+                        return false;
+                    }
+                }
+
+                // GroupClause is not supported by the conversion
+                return queryBody.SelectOrGroup is SelectClauseSyntax selectClause &&
+                    TryProcessSelectClause(selectClause, queryBody.Continuation, processingState);
```

Comment:
>queryBody.Continuation [](start = 57, length = 22)

Note that in continuation names of range variables from previous part of the query can be reused

Reply:
> I used names of those variables when attempted to construct GroupJoins. However, now I think that those variables can appear only in expressions that I pass as is. So, currently I do not see any actions I should apply to those variables. Do you see?

I think we can run into a name conflict situation. For example 
```
        foreach (var x in from x in new[] { 1 } select x + 2 into x where x > 0 select 7 into y let x = "aaa" select x )
        { }
```

---
In reply to: [177925979](https://github.com/dotnet/roslyn/pull/25362#discussion_r177925979) [](ancestors = 177925979,177887128)